ext/mini_racer_extension/mini_racer_extension.cc

```diff
@@ -128,11 +128,56 @@ typedef struct {
     size_t max_memory;
 } FunctionCall;
 
-enum IsolateFlags {
-    IN_GVL,
-    DO_TERMINATE,
-    MEM_SOFTLIMIT_VALUE,
-    MEM_SOFTLIMIT_REACHED,
+#define SLOT_BOUNDARY 1000 // every thousand denotes slot boundary
+
+class IsolateData {
+public:
+    enum Flag {
+        // first flags are bitfield
+        //  max count: sizeof(uintptr_t) * 8
+        IN_GVL, // whether we are inside of ruby gvl or not
+        DO_TERMINATE, // terminate as soon as possible
+        MEM_SOFTLIMIT_REACHED, // we've hit the memory soft limit
+
+        // softlimit size is full-sized uintptr_t
+        MEM_SOFTLIMIT_VALUE = 1 * SLOT_BOUNDARY,
+    };
+
+    static uintptr_t Get(Isolate *isolate, Flag flag) {
+        uintptr_t slotData, flagMask, retval;
+        uint slot = flag / SLOT_BOUNDARY;
+        slotData = reinterpret_cast<uintptr_t>(isolate->GetData(slot));
+
+        switch (slot) {
+        case 0:
+            flagMask = 1 << flag;
+            retval = (slotData & flagMask) == flagMask;
+            return retval;
+        case 1:
+            return slotData;
+        }
+
+        throw std::runtime_error("MINI_RACER_DEV_ERROR Flags didn't match switch statement in Get.");
```

Comment:
this is going to be the first instance of `throw` in the script, is it safe? should simply return "not set" for unknown slot?

Reply:
Really it's a dev-time error. I'd love for it to be a compile time error somehow -- there's no return value here that makes sense. (it's a uint, so -1 and 0 are both values to return that would be real bad and cause hard-to-debug issues)
We could assert (which is basically an exception as well) -- throwing is not 'safe' but should never be hit unless there's a bug in our implementation. 

I'm 100% open to ideas here. This was just the best I could come up with in the time I thought about it. :)

Reply:
as long as it is safe I guess I am fine with it, I guess it will cause a segfault if we mess this up?

Reply:
Gone in latest impl.